lib/Sema/CSSimplify.cpp

```diff
@@ -3313,6 +3313,53 @@ bool ConstraintSystem::repairFailures(
     return false;
   };
 
+  // Attempts to repair a mismatch where the srcType is a value of a nominal
+  // callable type which have a 'callAsFunction' that results in the dstType. So
+  // we can suggest make an explicit call.
+  auto repairByInsertingExplicitCallOnCallableType = [&](Type srcType,
+                                                         Type dstType) -> bool {
+    auto callAsFuncDecls = srcType->getCallAsFunctionDeclarations(DC);
```

Comment:
There is already a `DynamicCallableApplicableFnConstraint` I think we should use it instead of trying to match results manually here.

Reply:
Humm is the nominal with a `callAsFunction` the same as a `@dynamicCallable` type? But maybe we could use `ApplicableFnConstraint`? 
But in anyways I'm thinking this fix should be applicable to both `@dynamicCallable` and `callAsFunction` nominal type, so I have to change to handle both right I think

Reply:
Ugh sorry I totally messed one with another 🤦  But the general idea is the same, let's use `ApplicableFn` constraint to match this correctly

Reply:
That got me a bit confused too 😄 
But right ... let's try this, I'll ping you when I have something :)

Reply:
@xedin We have something, but it is not ready for review because I still have to figure out what happened with some diagnostics that regressed...

Reply:
For some reason, call for simplify...Constraint results in diagnostic regressions in cases like 
```
fSR13503(arg: valueg1) // expected-error {{cannot convert value of type 'SR13503A' to expected argument type 'String'}}
                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                           type of expression is ambiguous without more context
```
I still don't understand why it is returning `Solved` in those cases ... 
@xedin Am I missing something on this particular implementation?

Reply:
The problem is that you are using simplifyMemberConstraint which would simply lookup a `callAsFunction` members, that's not what we want here - we'd want to build `simplifyApplicableFnConstraint` which a fake `() -> <type var>` argument which would get applied to a `srcType`. `simplifyApplicableFnConstraint` should handle looking up of `callAsFunction` members and their use.

Reply:
Ahh thanks, do you mean something like
```
auto result = simplifyApplicableFnConstraint(
          FunctionType::get({}, dstType), srcType, /*trailingClosureMatching=*/None, TMF_ApplyingFix,
          getConstraintLocator(locator, LocatorPathElt::ApplyFunction()));
```
?

Reply:
Yeah, just be careful with locators.

Reply:
I remember trying that at first but it falls into some assertion with matching argument and getting argument info ... so I thought that a member constraint would make sense for this case :)

Reply:
The last commit uses `ApplicableFn` Had to relax some assertions and check for some arg label info(but maybe the ideal would be to teach `getArgumentInfo` to handle this locator... I'll check that)  But unfortunately this not resolve the initial problem ... but the good side is that since `ApplicableFn` does propagate to `DynamicCallableApplicableFn` so we would need just one check

Reply:
Seems like `simplifyApplicableFnConstraint` is always returning solved because it just adds the some a `ValueMember` and an `ApplicableFn` and always return solved https://github.com/apple/swift/blob/846f522b22318e8b0349e1c94cec574aaac135d8/lib/Sema/CSSimplify.cpp#L8842
I mistakenly thought that this would create the constraint attempt to solve it and return Solved/Error if matched/not ... but seems like it adds the constraint fails and actually doesn't record any diagnostics. And at the end, the explicit fix is always recorded because it always returns `Solved` ... 
So we actually have to think about how to solve this to use `simplifyApplicableFnConstraint`... 
Ahh and sorry for all the questions, but this behavior I was a bit surprising to me

Reply:
@xedin Am I getting this correctly? There is a way to check this and then without those side effects of the failed constraints? I've seen some examples of some checks like that being done by creating a new constraint system like in `TypeChecker::typesSatisfyConstraint` but not sure if this would work for this case ...

Reply:
Unfortunately I still couldn't think on how to implement this using `ApplicableFnConstraint` giving the things above 
So I reverted it and got back to lookup and matching approach to get everything right with test cases ... so if it's possible to implement this with the constraint I may need some hints to understand how to do it, can you give it a look when you have some time? :)

Reply:
Here is the problem - it's not always possible to determine whether contextual type would match result type of the `callAsFunction` overload directly in `repairFailures` because contextual type could be a generic parameter. 

Here is a hint at what needs to happen to diagnose this properly - you can make it work with or without using `simplifyApplicableFnConstraint` - constraint for each overload choice formed for an implicit call to `callAsFunction` variant has to have your fix attached so which would only recorded at the point when choice is applied. Also, there has to be a "fallback" case to a more generic fix e.g. "contextual mismatch" which mentions callable type directly, so if no overloads of `callAsFunction` resulted a valid solution it would be possible to diagnose a regular conversion mismatch directly on the type since we'd know that type is callable doesn't matter in that situation.

Reply:
Humm, thank you for the direction! I'll have to think a bit more to fully understand. I'll ping you for the review when I get this working :)

Reply:
Sounds good!

Reply:
Hey @xedin :) 
It took some time, but this should be ready for review